### PR TITLE
core: isolate X connection with error handling into a struct

### PR DESCRIPTION
**Pull request by yshui**
Part of the long running effort to reduce the prevalence of `session_t`. After this, functions that communicate with X can make use of the error handling machinary (set_ignore_cookie, set_cant_fail_cookie) without needing to take a `session_t` parameter.

This commit converts everything to use the new struct `x_connection`, most of the conversions are mechanical.

### CODE REVIEW

File: src/render.c [hunk 1]
```diff
@@ -162,26 +163,20 @@ static inline void attr_nonnull(1, 2) set_tgt_clip(session_t *ps, region_t *reg)
...
-	if (ppaint->pixmap)
-		xcb_free_pixmap(ps->c, ppaint->pixmap);
+	if (ppaint->pict != XCB_NONE) {
+		x_free_picture(&ps->c, ppaint->pict);
+	}
+	ppaint->pict = XCB_NONE;
+	if (ppaint->pixmap) {
+		xcb_free_pixmap(ps->c.c, ppaint->pixmap);
+	}
 	ppaint->pixmap = XCB_NONE;
```
Comment (absolutelynothelix) [inline src/render.c]:
```suggestion
	if (ppaint->pict != XCB_NONE) {
		x_free_picture(&ps->c, ppaint->pict);
		ppaint->pict = XCB_NONE;
	}
	if (ppaint->pixmap) {
		xcb_free_pixmap(ps->c.c, ppaint->pixmap);
		ppaint->pixmap = XCB_NONE;
	}
```

File: src/render.c [hunk 2]
```diff
@@ -1466,19 +1503,22 @@ void free_root_tile(session_t *ps) {
...
-		free_picture(ps->c, &ps->alpha_picts[i]);
+	for (int i = 0; i <= MAX_ALPHA; ++i) {
+		x_free_picture(&ps->c, ps->alpha_picts[i]);
+	}
 	free(ps->alpha_picts);
 	ps->alpha_picts = NULL;
 
 	// Free cshadow_picture and black_picture
-	if (ps->cshadow_picture == ps->black_picture)
+	if (ps->cshadow_picture == ps->black_picture) {
```
Comment (absolutelynothelix) [inline src/render.c]: silly github doesn't let me add a suggestion to proper lines
```suggestion
	if (ps->cshadow_picture != ps->black_picture) {
		x_free_picture(&ps->c, ps->cshadow_picture);
	}
	
	x_free_picture(&ps->c, ps->black_picture);
	x_free_picture(&ps->c, ps->white_picture);
	ps->cshadow_picture = ps->black_picture = ps->white_picture = XCB_NONE;
```

File: src/x.h [hunk 3]
```diff
@@ -103,15 +140,85 @@ static inline uint32_t x_new_id(xcb_connection_t *c) {
 	return ret;
 }
 
+static void set_reply_action(struct x_connection *c, uint32_t sequence,
+                             enum pending_reply_action action) {
+	auto i = cmalloc(pending_reply_t);
+	if (!i) {
+		abort();
+	}
```
Comment (absolutelynothelix) [inline src/x.h]: doesn't `cmalloc` check and report allocation failures itself?

File: src/x.h [hunk 4]
```diff
@@ -103,15 +140,85 @@ static inline uint32_t x_new_id(xcb_connection_t *c) {
...
+	c->pending_reply_tail = &c->pending_reply_head;
+
+	XSetErrorHandler(c->previous_xerror_handler);
+}
+
+/// Create a x_connection struct from a Xlib Display.
+///
+/// Note this function doesn't take ownership of the Display, user are still responsible
+/// for closing it after `free_x_connection` is called.
+void x_connection_init(struct x_connection *c, Display *dpy);
```
Comment (absolutelynothelix) [inline src/x.h]: * "a x_connection" => "an x_connection"
* "a Xlib" => "an Xlib"
* "user are" => "users are" or "user is"

File: src/x.h [hunk 5]
```diff
@@ -103,15 +140,85 @@ static inline uint32_t x_new_id(xcb_connection_t *c) {
...
+
+/// Discard queued pending replies.
+///
+/// We have received reply with sequence number `sequence`, which means all pending
+/// replies with sequence number less than `sequence` will ne1ver be received. So discard
```
Comment (absolutelynothelix) [inline src/x.h]: "ne1ver" => "never"

File: src/x.h [hunk 6]
```diff
@@ -309,7 +423,7 @@ xcb_screen_t *x_screen_of_display(xcb_connection_t *c, int screen);
  * ps->randr_monitor_regs with the data X RandR provided and the
  * x_free_randr_info function frees them.
  */
-void x_update_randr_monitors(session_t *ps);
-void x_free_randr_info(session_t *ps);
+void x_update_randr_monitors(struct x_connection *, struct x_monitors *);
+void x_free_monitor_info(struct x_monitors *);
```
Comment (absolutelynothelix) [inline src/x.h:419]: i'd choose simple and matching names, e.g. `x_update_monitors` and `x_free_monitors_info`. also, this makes the comment above outdated.

File: src/x.c [hunk 7]
```diff
@@ -28,6 +29,76 @@
...
+		if (sequence > c->pending_reply_head->sequence) {
+			auto next = c->pending_reply_head->next;
+			free(c->pending_reply_head);
+			c->pending_reply_head = next;
+			if (!c->pending_reply_head) {
+				c->pending_reply_tail = &c->pending_reply_head;
+			}
+		} else {
+			break;
+		}
```
Comment (absolutelynothelix) [inline src/x.c]:
```suggestion
		if (sequence <= c->pending_reply_head->sequence) {
			break;
		}

		auto next = c->pending_reply_head->next;
		free(c->pending_reply_head);
		c->pending_reply_head = next;
		if (!c->pending_reply_head) {
			c->pending_reply_tail = &c->pending_reply_head;
		}
```

File: src/x.c [hunk 8]
```diff
@@ -28,6 +29,76 @@
...
+		return;
+	}
+	x_log_error(LOG_LEVEL_WARN, ev->full_sequence, ev->major_code, ev->minor_code,
+	            ev->error_code);
+}
+
+/// Create a x_connection struct from a Xlib Display.
+///
+/// Note this function doesn't take ownership of the Display, user are still responsible
+/// for closing it after `free_x_connection` is called.
```
Comment (absolutelynothelix) [inline src/x.c]: same mistakes as in the `x.h`. btw, do we really need a duplicate comment?

File: src/x.c [hunk 9]
```diff
@@ -440,27 +511,39 @@ void x_set_picture_clip_region(xcb_connection_t *c, xcb_render_picture_t pict,
...
+/**
+ * Destroy a <code>Picture</code>.
+ *
+ * Picture must be valid.
+ */
+void x_free_picture(struct x_connection *c, xcb_render_picture_t p) {
+	assert(p != XCB_NONE);
+	auto cookie = xcb_render_free_picture(c->c, p);
+	set_cant_fail_cookie(c, cookie);
+}
```
Comment (absolutelynothelix) [inline src/x.c]: why do we care if a picture (or in some cases a pixmap) is successfully freed? and i don't see a particular logic when this function is preferred over direct `xcb_render_free_picture` call.

**Review by absolutelynothelix — COMMENTED** [review]
there are a lot of unrelated style and formatting changes. they should either be pushed directly and the pull request should be based on them or be a separate commit(s) because they make the actual reviewing harder and slower. the same applies to the unrelated logic changes. all of these changes are good but they should be separated to make reviewing easy. to prevent this in the future i suggest you to at least run clang-format on all .c and .h files and do a direct push.

overall, lgtm.

btw, a reminder. this todo seems to be related to these changes:
https://github.com/yshui/picom/blob/d4f72828f76ecacedfc2dcd0fb489f661be72bc8/src/x.c#L491

File: src/x.c (hunk 9 above)
Comment (yshui) [reply src/x.c]: this makes sure we don't double free a picture.

File: src/x.c (hunk 8 above)
Comment (yshui) [reply src/x.c]: yeah, this seems to be the convention

File: src/x.h (hunk 3 above)
Comment (yshui) [reply src/x.h]: yeah (don't look at me, i cut/pasted these)

File: src/x.c (hunk 9 above)
Comment (absolutelynothelix) [reply src/x.c]: it should be the x server's headache, not ours. furthermore, double freeing a picture pretty much results in an x error and if you want to trace where it happens there is [xrescheck](https://github.com/absolutelynothelix/xrescheck).

Comment (yshui) [reply src/x.c]: i don't quite understand what you are trying to say? `set_cant_fail_cookie` is what tracks the X error.

Comment (absolutelynothelix) [reply src/x.c]: calling the `xcb_render_free_picture` function pretty much sends a corresponding request to the x server and handling a possible double free is up to the x server. for us it's not something horrible like an actual double free in terms of memory allocated by us, it's just an x error. and i can't really imagine a scenario where something goes horribly wrong because of a double freed picture.

> `set_cant_fail_cookie` is what tracks the X error.

as i can see from the code, it considers the error unrecoverable and aborts.

basically, my point is that personally i wouldn't care if a picture is successfully freed or not because in most cases it'll result in a harmless error that could be traced with xrescheck. if you disagree with me i won't mind such mechanism but it's still unclear when to prefer it over a plain `xcb_render_free_picture` call because it doesn't seem to be always used.

Comment (yshui) [reply src/x.c]: because if we free a picture twice, it means there is a bug in our code, just like freeing a piece of memory twice. it would be nice if we can catch it.

it is completely irrelevant  if X handles this error internally or not.

does that make sense?

Comment (absolutelynothelix) [reply src/x.c]: these errors are not that severe to abort on them and could be caught without this mechanism but... nevermind. i'd like to see other nitpicks addressed and i'll merge this pull request.

Comment (yshui) [reply src/x.c]: well there is also `set_debug_cant_fail_cookie`

**Comment by yshui** [comment]
@absolutelynothelix I've fixed all your points. thanks for the review btw.

> btw, a reminder. this todo seems to be related to these changes:

I think this comment is referring to the `ps->*_error` error codes.

**Comment by absolutelynothelix** [comment]
> I think this comment is referring to the `ps->*_error` error codes.

yeah, i think you wanted to separate the `*_error` (`*_event` and `has_*` too, probably) members from the `session_t` structure. i just found this todo relevant to the current changes and decided to remind you about it so you may take it if you find it relevant too.
